sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala

```diff
@@ -1059,6 +1060,72 @@ class FileStreamSourceSuite extends FileStreamSourceTest {
     val str = Source.fromFile(getClass.getResource(s"/structured-streaming/$file").toURI).mkString
     SerializedOffset(str.trim)
   }
+
+  test("FileStreamSource - latestFirst") {
+    withTempDir { src =>
+      // Prepare two files: 1.txt, 2.txt, and make sure they have different modified time.
+      val f1 = stringToFile(new File(src, "1.txt"), "1")
+      val f2 = stringToFile(new File(src, "2.txt"), "2")
+      eventually(timeout(streamingTimeout)) {
+        f2.setLastModified(System.currentTimeMillis())
+        assert(f1.lastModified < f2.lastModified)
+      }
+
+      // Read oldest files first, so the first batch is "1", and the second batch is "2".
+      val fileStream = createFileStream(
+        "text",
+        src.getCanonicalPath,
+        options = Map("latestFirst" -> "false", "maxFilesPerTrigger" -> "1"))
+      val clock = new StreamManualClock()
+      testStream(fileStream)(
+        StartStream(trigger = ProcessingTime(10), triggerClock = clock),
+        AssertOnQuery { _ =>
+          // Block until the first batch finishes.
+          eventually(timeout(streamingTimeout)) {
+            assert(clock.isStreamWaitingAt(0))
+          }
+          true
+        },
+        CheckLastBatch("1"),
+        AdvanceManualClock(10),
+        AssertOnQuery { _ =>
+          // Block until the second batch finishes.
+          eventually(timeout(streamingTimeout)) {
+            assert(clock.isStreamWaitingAt(10))
+          }
+          true
+        },
+        CheckLastBatch("2")
+      )
+
+      // Read latest files first, so the first batch is "2", and the second batch is "1".
+      val fileStream2 = createFileStream(
```

Comment:
I think this code can be deduped by writing a function that make the query run two batches and collect the results in order. And then the function is called with latestFirst true or false, and the result order checked.